src/System.Linq/src/System/Linq/Reverse.cs

```diff
@@ -20,7 +20,7 @@ public static IEnumerable<TSource> Reverse<TSource>(this IEnumerable<TSource> so
             return new ReverseIterator<TSource>(source);
         }
 
-        private sealed class ReverseIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
+        private sealed class ReverseIterator<TSource> : Iterator<TSource>, IPartition<TSource>
```

Comment:
@stephentoub The caller may not always have `source` cached in a local, so it may be more convenient to just get the first from `Reverse()` rather then reevaluating `source`. For example:

```cs
var items = cus.Where(c => c.Id == x).Reverse();

items.First(); // As opposed to calling cus.Last(c => c.Id == x)
items.Last(); // As opposed to calling cus.First(c => c.Id == x)
```

Another example is that `source.Reverse()` could have been passed to another method where `First` or `Last` is called on it.

An additional reason for making these changes was that they were pretty trivial-- there are more red diffs than green in this PR in `src/`. And the benefits are pretty substantial; we are not just eliminating virtual method calls, we are also reducing allocations/big O complexity.